compiler/luci/service/src/Nodes/CircleFullyConnected.cpp

```diff
@@ -36,4 +40,76 @@ luci::CircleNode *CloneNodeLet<CN::DEF>::visit(const luci::CircleFullyConnected
   return cloned;
 }
 
+namespace sinf
+{
+
+loco::TensorShape Algorithm::visit(const luci::CircleFullyConnected *node)
+{
+  auto input_shape = circle_shape(loco::must_cast<CircleNode *>(node->input()));
+  auto weights_shape = circle_shape(loco::must_cast<CircleNode *>(node->weights()));
+
+  loco::TensorShape out_shape;
+
+  // NOTE Some recipes in some repositories are using rank 4 input for FullyConnected.
+  //      Until they are all fixed, disable following assert.
+  // TODO Enable following assert after related fixes are applied
+  // https://github.com/tensorflow/tensorflow/blob/ea33c1e7a25d8025e8ee405ad8ab7be261798d76/tensorflow/lite/kernels/fully_connected.cc#L194
+  // LUCI_ASSERT(input_shape.rank() == 2 || input_shape.rank() == 3,
+  //             "Input rank of FullyConnected should be 2 or 3");
+
+  // https://github.com/tensorflow/tensorflow/blob/ea33c1e7a25d8025e8ee405ad8ab7be261798d76/tensorflow/lite/kernels/fully_connected.cc#L225
+  LUCI_ASSERT(weights_shape.rank() == 2, "Weights of FullyConnected should be 2");
+
+  // https://github.com/tensorflow/tensorflow/blob/ea33c1e7a25d8025e8ee405ad8ab7be261798d76/tensorflow/lite/kernels/fully_connected.cc#L353-L367
+  if (node->keep_num_dims())
+  {
+    // [1,10,?,20] X [15,20] -> [1,10,?,15] (weights_shape =[15,20])
+    out_shape.rank(input_shape.rank());
+    for (uint32_t i = 0; i < input_shape.rank(); ++i)
+      out_shape.dim(i) = input_shape.dim(i);
+    out_shape.dim(out_shape.rank() - 1) = weights_shape.dim(0);
+  }
+  else
+  {
+    bool is_dynamic_shape = false;
+
+    for (uint32_t i = 0; i < input_shape.rank() - 1; i++)
+    {
+      if (not input_shape.dim(i).known())
+      {
+        is_dynamic_shape = true;
+        break;
+      }
+    }
+
+    // Originally, BatchSize = input_size / weights_shape(0).value()
```

Comment:
what does it mean by "Originally"?

can you give some reference about `BatchSize = input_size / weights_shape(0).value()`?
I don't understand this equation.

Reply:
variable used below is `batch_size` but you used `BatchSize`.
it would be better to use one notation.

Reply:
> what does it mean by "Originally"?
> 
> can you give some reference about `BatchSize = input_size / weights_shape(0).value()`? I don't understand this equation.

"Originally" refers to the code written before any modifications.
(compiler/luci/service/src/CircleShapeInferenceRule.cpp#L714-L718)

According to the link,(https://github.com/tensorflow/tensorflow/blob/ea33c1e7a25d8025e8ee405ad8ab7be261798d76/tensorflow/lite/kernels/fully_connected.cc#L218-L228), when TensorFlow performs the FullyConnected operation, it first determines the input_size and then decides the batch_size.
However, the value of the last dimension of input_shape, which is multiplied at the end, is the same as the value of the first element of weights_shape.

Reply:
> can you give some reference about `BatchSize = input_size / weights_shape(0).value()`? I don't understand this equation.

Sorry, I made a mistake, it should be 'batch_size = input_size / weights_shape(1).value().'

Reply:
It is not necessary to include comments in your code explaining the differences from the code before the modification.

However, the two sections of code below seem to have been modified.
It would be helpful to have an explanation of whether these changes were made to address a specific 'issue' or if they were simply intended to simplify the code.

### your code 

```c++
    uint32_t input_batch = 1;

    for (uint32_t i = 0; i < input_shape.rank() - 1; i++)
    {
      input_batch *= input_shape.dim(i).value();
    }
```

### original code

```c++
    uint32_t input_size = 1;
    for (uint32_t i = 0; i < input_shape.rank(); i++)
    {
      input_size = input_size * input_shape.dim(i).value();
    }
    const uint32_t batch_size = input_size / weights_shape.dim(1).value();
```

Reply:
I modified the code to prevent a specific case.

In the calculation, input_size is determined by multiplying the values from 0 to rank()-1, and then the batch size is calculated by dividing this value by the last dimension's value of weights_shape.

Since in the FullyConnected operation, the value of the last dimension of weights_shape is the same as the value of the last dimension of input_shape, the result is effectively the same as multiplying up to rank()-2.

The issue arises if input_size is calculated by multiplying from 0 to rank()-1 and the last dimension of input_shape is dynamic. Since the value of the dynamic shape is set to 0, the input_size becomes 0, resulting in a batch_size of 0 even if it should not be.

Reply:
> and the last dimension of input_shape is dynamic.

is this the only case for input having dynamic shape? 
what happens if input is `[?, ?, ?]` ?

Reply:
> > and the last dimension of input_shape is dynamic.
> 
> is this the only case for input having dynamic shape? what happens if input is `[?, ?, ?]` ?

I separated checking dynamic shape part and calculating batch_size part.
As I know, the shape of the weights is determined by the size of the input tensor and is finalized when the model is defined.

Therefore, when performing the matrix multiplication between the input and the weights, the last dimension of the output will be the value of the first dimension of the weight matrix.

Result : [?, weight_shape.dim(1).value()]